### PR TITLE
Admit critical pods in the kubelet

### PR DESCRIPTION
**Pull request by bprashanth**
Teaches the node about critical pods and makes kube-proxy critical. 
Ref https://github.com/kubernetes/kubernetes/issues/38322

### CODE REVIEW

File: cluster/saltbase/salt/kube-proxy/kube-proxy.manifest [hunk 1]
```diff
@@ -41,6 +41,12 @@ metadata:
   labels:
     tier: node
     component: kube-proxy
+    # Even though kube-proxy runs as a static pod, this annotation lowers the
```
Comment (davidopp) [inline cluster/saltbase/salt/kube-proxy/kube-proxy.manifest]: I would remove the "Even though kube-proxy runs as a static pod" part, and then add a new sentence at the end "Note that kube-proxy runs as a static pod so this annotation does NOT have any effect on rescheduler (default scheduler and rescheduler are not involved in scheduling kube-proxy)."

**Comment by davidopp** [comment]
Just to be clear, by reusing the critical-pod annotation, you are giving this new Kubelet behavior to a bunch of pods (that currently have the annotation), such as Heapster, DNS, and the kube dashboard. Are you OK with this?

**Comment by bprashanth** [comment]
Basically we're saying we want these pods to: 
* always get scheduled
* always get admitted
* not get oom killed

regardless of what resource/request specified. 

If there's resource pressure, some other non-critical pod will get "disrupted". This disruption might happen at the scheduler level (normal case) or at admission time on the node. 

Previously, the following scenario: 
* scheduler placed best-effort critical pod on node 
* node is under memory pressure because of some system daemon (so scheduler is not aware)
* best-effort critical pod is not admitted, guaranteed user pods run on that node instead

might now become:
* scheduler placed best-effort critical pod on node (or kubelet finds a best effort critical static pod)
* node is under memory pressure because of some system daemon (so scheduler is not aware)
* user guaranteed pod is not run, instead we run the best-effort critical pod

So now a guaranteed pod might get "disrupted", but the critical bit is supposed to actually override QoS (i.e critical pod ties are broked on QoS, is how I imagine it working). Does that make sense?

**Comment by thockin** [comment]
I am ill-equipped to review this.  @dchen1107 can you assign a reviewer?

**Comment by davidopp** [comment]
@bprashanth I understand. All I am saying is that by reusing the critical-pod annotation, you are "grandfathering" a lot of pods that already use the annotation (at least Heapster, DNS, and dashboard) into this new behavior. If that is the intent, this is fine, I just wanted to mention it.

**Comment by bprashanth** [comment]
Since we're targetting a point release, maybe it's better to split them out? thinking through the cases

**Comment by dchen1107** [comment]
I agreed with @davidopp above, on a given node, kube-proxy is even more critical than Heapster, DNS, and the kube dashboard, which can be moved to a different node at the cluster level. In a long run, when we have a cluster-level eviction, we still need to distinguish kube-proxy like critical pod from other critical ones. I am ok by introducing a new annotation for now.

**Comment by dchen1107** [comment]
Re: https://github.com/kubernetes/kubernetes/pull/38836#issuecomment-267472708 

Again, I agreed with @davidopp there are two different levels of the critical and importance. And I admit there is a rare case when heapster, DNS, dashboard, etc. cluster addons (also critical pods) are all scheduled on the same node, and the node runs into the resource starvation issue, and kube-proxy might still be evicted here.  

But @bprashanth just pointed out to me that except DNS, other cluster-level critical pods are already configured with guaranteed qos (oom_score_adj: -998). In this case, introducing another annotation doesn't help much as a patch. To simplify our patch, can we just introduce one annotation for now, and come back with fresh mind next year for #22212? @davidopp WDYT here?

File: cluster/saltbase/salt/kube-proxy/kube-proxy.manifest [hunk 2]
```diff
@@ -38,6 +38,14 @@ kind: Pod
 metadata:
   name: kube-proxy
   namespace: kube-system
+  # This annotation lowers the possibility that kube-proxy gets evicted when the
+  # node is under memory pressure, and prioritizes it for admission, even if
```
Comment (dchen1107) [inline cluster/saltbase/salt/kube-proxy/kube-proxy.manifest]: Can you change the memory pressure to the resource pressure / starvation? The node eviction manager works for both memory and disk today.

Comment (bprashanth) [inline cluster/saltbase/salt/kube-proxy/kube-proxy.manifest]: Hmm, I don't think anything in the kubelet reads the disk pressure node condition, does it? 

if we're out of disk we will still not admit it (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1677) and if we admit it in that case it'll probably make the disk situation worse? 

The current fix is only around the admit method of the eviction manager, and only based on the NodeMemoryPressure condition.

Comment (mtaufen) [inline cluster/saltbase/salt/kube-proxy/kube-proxy.manifest]: This implementation looks like it just considers critical pods for memory pressure, though.

**Comment by bprashanth** [comment]
the good thing about using a single annotation is that DNS will get the described behavior (it is the only addon that's not guaranteed but we want -999 oom score).

File: pkg/kubelet/eviction/eviction_manager_test.go [hunk 3]
```diff
@@ -210,6 +211,8 @@ func TestMemoryPressure(t *testing.T) {
 	// create a best effort pod to test admission
 	bestEffortPodToAdmit, _ := podMaker("best-admit", newResourceList("", ""), newResourceList("", ""), "0Gi")
 	burstablePodToAdmit, _ := podMaker("burst-admit", newResourceList("100m", "100Mi"), newResourceList("200m", "200Mi"), "0Gi")
+	criticalBestEffortPodToAdmit, _ := podMaker("critial-best-admin", newResourceList("", ""), newResourceList("", ""), "0Gi")
```
Comment (mtaufen) [inline pkg/kubelet/eviction/eviction_manager_test.go]: s/critical-best-admin/critial-best-admit

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -1910,6 +1910,10 @@ func (kl *Kubelet) handleMirrorPod(mirrorPod *v1.Pod, start time.Time) {
 func (kl *Kubelet) HandlePodAdditions(pods []*v1.Pod) {
 	start := kl.clock.Now()
 	sort.Sort(sliceutils.PodsByCreationTime(pods))
+
+	// Pass critical pods through admission check first.
+	sort.Sort(kubepod.SortPrioritizingCriticalPods(pods))
```
Comment (mtaufen) [inline pkg/kubelet/kubelet.go]: Do we still want the critical/non-critical pods to be sub-sorted by their creation time? I don't think `sort.Sort` guarantees a stable sort.

File: pkg/kubelet/kubelet_test.go [hunk 5]
```diff
@@ -461,6 +461,62 @@ func TestHandlePortConflicts(t *testing.T) {
 	require.Equal(t, v1.PodPending, status.Phase)
 }
 
+// Tests that we sort pods based on criticality.
+func TestCriticalPrioritySorting(t *testing.T) {
```
Comment (mtaufen) [inline pkg/kubelet/kubelet_test.go]: If it turns out that we *do* want a criticality sort that is stable wrt creation time, we should check that in this test.

File: pkg/kubelet/types/pod_update.go [hunk 6]
```diff
@@ -27,6 +27,15 @@ const ConfigMirrorAnnotationKey = "kubernetes.io/config.mirror"
 const ConfigFirstSeenAnnotationKey = "kubernetes.io/config.seen"
 const ConfigHashAnnotationKey = "kubernetes.io/config.hash"
 
+// This key needs to sync with the key used by the rescheduler, which currently
+// lives in contrib. Its presence indicates 2 things, as far as the kubelet is
+// concerned:
+// 1. Resource related admission checks will prioritize the admission of
+//    pods bearing the key, over pods without the key, regardless of QoS.
+// 2. The OOM score of pods bearing the key will be <= pods without
```
Comment (mtaufen) [inline pkg/kubelet/types/pod_update.go]: I don't see any adjustments to OOM score based on criticality in this PR, where is that taking place?

File: cluster/saltbase/salt/kube-proxy/kube-proxy.manifest (hunk 2 above)
Comment (mtaufen) [inline cluster/saltbase/salt/kube-proxy/kube-proxy.manifest]: I agree with @bprashanth, let's focus on the memory situation first.

File: pkg/kubelet/types/pod_update.go (hunk 6 above)
Comment (bprashanth) [inline pkg/kubelet/types/pod_update.go]: not in this pr

**Comment by davidopp** [comment]
I'm fine either way.

File: cluster/saltbase/salt/kube-proxy/kube-proxy.manifest (hunk 2 above)
Comment (dchen1107) [inline cluster/saltbase/salt/kube-proxy/kube-proxy.manifest]: @bprashanth Node eviction manager take actions on both memory and disk (space and inode). I know your current implementation only consider memory pressure condition, not include disk. I am ok as is here.

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (bprashanth) [inline pkg/kubelet/kubelet.go]: yeah fair point, ill add something

**Comment by dchen1107** [comment]
lgtm. @davidopp @mtaufen any other concerns?

**Comment by mtaufen** [comment]
this lgtm

**Comment by bprashanth** [comment]
LGTM label anyone?

**Comment by bprashanth** [comment]
oh, didn't realize I had 2 verbal confirmations, applying

**Comment by bprashanth** [comment]
I think we should also reduce the chances of a critical pod getting evicted (this pr only deals with admission): https://github.com/kubernetes/kubernetes/pull/38914

**Comment by vishh** [comment]
Why are we allowing "best effort" pods to become critical?

**Comment by smarterclayton** [comment]
So this basically bypasses all enforced safety on the platform - with this change, anyone can DOS all nodes even if quota and limit range is in place and nodeName is defended against.  Can we get this put behind a flag, so tenants can't trivially consume all resources on the platform?

**Comment by smarterclayton** [comment]
@derekwaynecarr  @pweil- we may have to restrict access to this annotation in an admission controller to someone who can direct bind to nodes - we should make sure we don't pick this up in back ports until we do a security vetting on it or patch it out.

**Comment by bprashanth** [comment]
What you described was already happening on account of poor disk eviction, and it's what led to this set of changes to begin with. Basically some random best effort pod would dos the node through disk (logs, writing to container local fs) bouncing all pods, guaranteed, static or otherwise. 

What made the situation worse was we had logic to prevent admission of best-effort pods under memory pressure, even if that pod had nothing to do with the pressure. All this pr does is lift that restriction (https://github.com/kubernetes/kubernetes/pull/38836/files#diff-655a99a375f3331c5e98e9f4a6f0bd96R112) by examining the annotation.
 
Dawn had another pr that set the oom score. That one is admittedly more dangerous. I'd argued initially that we should just do it for static pods instead of through the "critical" annotation. 

Of course this is everyone's problem, right from not setting kube-proxy default limits to improper eviction and coming up with a better priority scheme (https://github.com/kubernetes/kubernetes/issues/39124). I think we just need to prioritize all fixes we've been delaying ASAP, users (and by extension, on-call) are starting to feel the pain.

**Comment by pweil-** [comment]
> @derekwaynecarr @pweil- we may have to restrict access to this annotation in an admission controller to someone who can direct bind to nodes - we should make sure we don't pick this up in back ports until we do a security vetting on it or patch it out.

@soltysh - please check for this in your rebase

**Comment by derekwaynecarr** [comment]
@bprashanth -- i do not agree with your assessment.

**Pull request merged**